src/cpp/core/system/Win32Pty.cpp

```diff
@@ -0,0 +1,578 @@
+                                   NULL /*lpSecurityAttributed*/,
+                                   OPEN_EXISTING,
+                                   0 /*dwFlagsAndAttributes*/,
+                                   NULL /*hTemplateFile*/);
+      if (*pStdInWrite == INVALID_HANDLE_VALUE)
+      {
+         DWORD err = ::GetLastError();
+         stopPty();
+         *pStdInWrite = NULL;
+         ::SetLastError(err);
+         return systemError(::GetLastError(),
+                            "Failed to connect to pty conin pipe",
+                            ERROR_LOCATION);
+      }
+   }
+
+   if (pStdOutRead && conout_name && conout_name(pPty_))
+   {
+      *pStdOutRead = ::CreateFileW(conout_name(pPty_),
+                                   GENERIC_READ,
+                                   0 /*dwShareMode*/,
+                                   NULL /*lpSecurityAttributed*/,
+                                   OPEN_EXISTING,
+                                   0 /*dwFlagsAndAttributes*/,
+                                   NULL /*hTemplateFile*/);
+      if (*pStdOutRead == INVALID_HANDLE_VALUE)
+      {
+         DWORD err = ::GetLastError();
+         stopPty();
+         *pStdOutRead = NULL;
+         return systemError(err,
+                            "Failed to connect to pty conout pipe",
+                            ERROR_LOCATION);
+      }
+   }
+
+   if (pStdErrRead && conerr_name && conerr_name(pPty_))
+   {
+      *pStdErrRead = ::CreateFileW(conerr_name(pPty_),
+                                   GENERIC_READ,
+                                   0 /*dwShareMode*/,
+                                   NULL /*lpSecurityAttributed*/,
+                                   OPEN_EXISTING,
+                                   0 /*dwFlagsAndAttributes*/,
+                                   NULL /*hTemplateFile*/);
+      if (*pStdOutRead == INVALID_HANDLE_VALUE)
+      {
+         DWORD err = ::GetLastError();
+         stopPty();
+         *pStdErrRead = NULL;
+         return systemError(err,
+                            "Failed to connect to pty conerr pipe",
+                            ERROR_LOCATION);
+      }
+   }
+
+   closeStdInWrite.detach();
+   closeStdOutRead.detach();
+   closeStdErrRead.detach();
+   return Success();
+}
+
+Error WinPty::runProcess(HANDLE* pProcess)
+{
+   if (pProcess)
+      *pProcess = NULL;
+
+   if (!ptyRunning())
+   {
+      return systemError(boost::system::errc::no_such_process,
+                         "Pty not running",
+                         ERROR_LOCATION);
+   }
+
+   // process command line arguments (copy of approach done by non-pseudoterm
+   // code path in ChildProcess::run for Win32)
+   std::string cmdLine;
+   BOOST_FOREACH(std::string& arg, args_)
+   {
+      cmdLine.push_back(' ');
+
+      // This is kind of gross. Ideally we would be more deterministic
+      // than this.
+      bool quot = std::string::npos != arg.find(' ')
```

Comment:
Quoted arguments on Windows is still a little messy; see e.g. here: https://github.com/rstudio/rstudio/blob/d13a0b9d4cba8e44b7beee0b8f74c8b611bbe3cf/src/cpp/session/SessionAsyncRProcess.cpp#L99-L111